h2/src/main/org/h2/command/dml/Merge.java

```diff
@@ -338,13 +339,10 @@ public boolean isCacheable() {
         return true;
     }
 
-    public TableFilter getTableFilter() {
-        return tableFilter;
-    }
-
-    public void setTableFilter(TableFilter tableFilter) {
-        this.tableFilter = tableFilter;
-        setTable(tableFilter.getTable());
+    @Override
+    public void collectDependecies(HashSet<DbObject> dependencies) {
+        if (query != null) {
+            query.collectDependecies(dependencies);
+        }
```

Comment:
What about `table` / `update` (`update` may be null)? They don't need to be collected, because `update` either find something to update, either don't?

Reply:
I think they don't. But `valuesExpressionList` may need it, because it may contain subqueries.

Reply:
And not only the `Merge`, the plain `Insert` and `Update` also may have such things in their list of values.

Reply:
I had them initially, but it fails to work, because it won;t see it's own changes. The whole thing is rather hacky. I think each TableFilter should be qualified as "read" (source) or "update" (target). Depends on this related cursors should be taken either on a snapshot or current map. Now if table makes it into dependencies, cursor will be based on a snapshot, and that's not what we want for merge target TableFilter.
As far as valuesExpressionList goes - I am not sure how it suppose to be. On one hand, there could be some queries, on the other hand - use of outdated values in selection list is counter intuitive, especially if list is used for insert/update/merge.

Reply:
Hypothetically the whole insert or set list should be consistent with itself that is not ensured by H2 with and without these changes.

If its processing here leads to failures I think you can merge this PR as is.